pkg/ccl/testccl/sqlccl/sqlccl_test: TestExplainRedactDDL: EXPLAIN (OPT, REDACT) CREATE FUNCTION does not redact

pkg/ccl/testccl/sqlccl/sqlccl_test.TestExplainRedactDDL [failed](https://tanzanite.cluster.engflow.com/invocations/default/fc105e6f-638e-4c6e-9904-2b8c9034a9f3?testReportRun=16&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL3Rlc3RjY2wvc3FsY2NsOnNxbGNjbF90ZXN0) on release-24.2.0-rc @ [46c11ab2cf6fa6f6adf7abb99ef72f68e4d538f8](https://github.com/cockroachdb/cockroach/commits/46c11ab2cf6fa6f6adf7abb99ef72f68e4d538f8):


```
        	NOT LEAKPROOF
        	AS $funcbody$
        INSERT INTO defaultdb.public."tAbl\\U00031706e_2" AS tab_2(col2_6, col2_5, "c%27ol2_4", "c\\x54ol}2_3", "Col2_2", """\rco😣l2_1") VALUES ('0101000060E6100000DF1D1A1DE00856C0847F418D157F4CC04CE89A0F3B17FC41':::GEOGRAPHY, '\xd89c':::BYTES, e'\U00053ADC\U0007AA5B\U00067165\U00002521' COLLATE de_DE, 2.084592822529178063E+36:::DECIMAL, (-3.16690563829366663):::DECIMAL, '02:15:48.063916':::TIME);
        INSERT INTO defaultdb.public."'table_3" AS tab_3("{c""ol3_0", col3_3, "c
        o{l3_2", col3_1) VALUES ('3effeaff-9afd-49ba-9919-9735c4009767':::UUID, B'010110001110010000111101110111011000110101', '{"bar": null, "pKdIRlT4z": "Coh35"}':::JSONB, (-1):::INT8);
        SELECT (-3.335520063081170484E+36):::DECIMAL AS col_1, (-322382181):::INT8 AS col_2;
        UPDATE defaultdb.public.table_1 AS tab_4 SET col1_6 = tab_4.col1_6 FROM defaultdb.public."'table_3" AS tab_5, defaultdb.public.table_1 AS tab_6, defaultdb.public."tAbl\\U00031706e_2" AS tab_7, defaultdb.public.table_1 AS tab_8 WHERE false LIMIT 92:::INT8;
        SELECT tab_9." c😚ol 1_1" AS col_3, tab_9.tableoid AS col_4, 'pachycephalosaurus8^ZB':::STRING AS col_5 FROM defaultdb.public.table_1 AS tab_9 WHERE NULL ORDER BY tab_9."	col1_7";
        UPDATE defaultdb.public.table_1 AS tab_10 SET "😴co l1_0" = tab_10."😴co l1_0" ORDER BY tab_10."�c/ol1_14" DESC LIMIT 35:::INT8;
        SELECT count(*) AS col_6 FROM defaultdb.public.table_1 AS tab_11 GROUP BY tab_11.col1_2 LIMIT 70:::INT8;
        SELECT p0 AS col_7, (-23802):::INT8 AS col_8, tab_13.col3_3 AS col_9, tab_13.col3_4 AS col_10 FROM defaultdb.public."tAbl\\U00031706e_2" AS tab_12 JOIN defaultdb.public."'table_3"@"'table_3_pkey" AS tab_13 ON (tab_12.col2_7) = (tab_13.crdb_internal_mvcc_timestamp) ORDER BY tab_12.col2_5 NULLS FIRST LIMIT 36:::INT8;
        SELECT NULL AS col_11 FROM defaultdb.public."'table_3" AS tab_14
        $funcbody$
        output:
        create-function
         ├── CREATE OR REPLACE FUNCTION func_1(INOUT p0 DECIMAL, OUT p1 INT4)
         │   	RETURNS RECORD
         │   	LANGUAGE SQL
         │   	VOLATILE
         │   	NOT LEAKPROOF
         │   	AS $$INSERT INTO defaultdb.public."tAbl\\U00031706e_2" AS tab_2(col2_6, col2_5, "c%27ol2_4", "c\\x54ol}2_3", "Col2_2", """\rco😣l2_1") VALUES ('0101000060E6100000DF1D1A1DE00856C0847F418D157F4CC04CE89A0F3B17FC41':::GEOGRAPHY, e'\\xd89c':::BYTES, e'\U00053ADC\U0007AA5B\U00067165\U00002521':::STRING COLLATE de_DE, 2.084592822529178063E+36:::DECIMAL, (-3.16690563829366663):::DECIMAL, '02:15:48.063916':::TIME);
         │   INSERT INTO defaultdb.public."'table_3" AS tab_3("{c""ol3_0", col3_3, "c
         │   o{l3_2", col3_1) VALUES ('3effeaff-9afd-49ba-9919-9735c4009767':::UUID, B'010110001110010000111101110111011000110101', '{"bar": null, "pKdIRlT4z": "Coh35"}':::JSONB, (-1):::INT8);
         │   SELECT (-3.335520063081170484E+36):::DECIMAL AS col_1, (-322382181):::INT8 AS col_2;
         │   UPDATE defaultdb.public.table_1 AS tab_4 SET col1_6 = tab_4.col1_6 FROM defaultdb.public."'table_3" AS tab_5, defaultdb.public.table_1 AS tab_6, defaultdb.public."tAbl\\U00031706e_2" AS tab_7, defaultdb.public.table_1 AS tab_8 WHERE false LIMIT 92:::INT8;
         │   SELECT tab_9." c😚ol 1_1" AS col_3, tab_9.tableoid AS col_4, 'pachycephalosaurus8^ZB':::STRING AS col_5 FROM defaultdb.public.table_1 AS tab_9 WHERE NULL ORDER BY tab_9."	col1_7";
         │   UPDATE defaultdb.public.table_1 AS tab_10 SET "😴co l1_0" = tab_10."😴co l1_0" ORDER BY tab_10."�c/ol1_14" DESC LIMIT 35:::INT8;
         │   SELECT count(*) AS col_6 FROM defaultdb.public.table_1 AS tab_11 GROUP BY tab_11.col1_2 LIMIT 70:::INT8;
         │   SELECT p0 AS col_7, (-23802):::INT8 AS col_8, tab_13.col3_3 AS col_9, tab_13.col3_4 AS col_10 FROM defaultdb.public."tAbl\\U00031706e_2" AS tab_12 JOIN defaultdb.public."'table_3"@"'table_3_pkey" AS tab_13 ON (tab_12.col2_7) = (tab_13.crdb_internal_mvcc_timestamp) ORDER BY tab_12.col2_5 NULLS FIRST LIMIT 36:::INT8;
         │   SELECT NULL AS col_11 FROM defaultdb.public."'table_3" AS tab_14;$$
         ├── dependencies
         │    ├── "tAbl\\U00031706e_2" [columns: """\rco😣l2_1" "Col2_2" "c%27ol2_4" "c\\x54ol}2_3" col2_5 col2_6 col2_7]
         │    ├── "'table_3" [columns: "c
         │    │   o{l3_2" "{c""ol3_0" col3_1 col3_3 col3_4]
         │    ├── table_1 [columns: "�c/ol1_14" " c😚ol 1_1" "%b1col1_9" "c%vol1 _10" "co l1(_12" "col
1😦_3" "col 1_16" col1_13 col1_15 col1_2 col1_5 col1_6]
         │    ├── "'table_3" [no columns]
         │    ├── table_1 [no columns]
         │    ├── "tAbl\\U00031706e_2" [no columns]
         │    ├── table_1 [no columns]
         │    ├── table_1 [columns: "	col1_7" " c😚ol 1_1" tableoid]
         │    ├── table_1 [columns: "�c/ol1_14" " c😚ol 1_1" "%b1col1_9" "c%vol1 _10" "co l1(_12" "col
1😦_3" "col 1_16" "😴co l1_0" col1_13 col1_15 col1_2 col1_5]
         │    ├── table_1 [columns: col1_2]
         │    ├── "tAbl\\U00031706e_2" [columns: col2_5 col2_7]
         │    ├── "'table_3"@'table_3_pkey [columns: col3_3 col3_4 crdb_internal_mvcc_timestamp]
         │    └── "'table_3" [no columns]
         ├── cardinality: [0 - 0]
         ├── volatile, mutations
         ├── stats: [rows=0]
         └── cost: 0.01
            --- FAIL: TestExplainRedactDDL/Explain/Opt/Verbose (0.04s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=16</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #121545 pkg/ccl/testccl/sqlccl/sqlccl_test: TestExplainRedactDDL failed [C-test-failure O-robot P-2 T-sql-queries branch-release-23.1.18-rc]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestExplainRedactDDL.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-40913

--- a/pkg/ccl/logictestccl/testdata/logic_test/explain_redact
+++ b/pkg/ccl/logictestccl/testdata/logic_test/explain_redact
@@ -338,3 +338,68 @@ memo (optimized, ~7KB, required=[presentation: info:5] [distribution: test])
  └── G7: (const ‹×›)
 scan q
  └── constraint: /1: ‹×›
+
+# Regression test for #128282: check EXPLAIN (OPT, REDACT) of various CREATE
+# statements.
+
+statement ok
+CREATE TABLE t128282 (col STRING)
+
+query T
+EXPLAIN (OPT, REDACT) CREATE FUNCTION f() RETURNS STRING LANGUAGE SQL AS $$ SELECT * FROM t128282 WHERE col = 'secret' $$
+----
+create-function
+ ├── CREATE FUNCTION f()
+ │     RETURNS STRING
+ │     LANGUAGE SQL
+ │     AS $$SELECT t128282.col FROM test.public.t128282 WHERE col = ‹×›;$$
+ └── dependencies
+      └── t128282 [columns: col]
+
+query T
+EXPLAIN (OPT, VERBOSE, REDACT) CREATE FUNCTION f() RETURNS STRING LANGUAGE SQL AS $$ SELECT * FROM t128282 WHERE col = 'secret' $$
+----
+create-function
+ ├── CREATE FUNCTION f()
+ │     RETURNS STRING
+ │     LANGUAGE SQL
+ │     AS $$SELECT t128282.col FROM test.public.t128282 WHERE col = ‹×›;$$
+ ├── dependencies
+ │    └── t128282 [columns: col]
+ ├── cardinality: [0 - 0]
+ ├── volatile, mutations
+ ├── stats: [rows=0]
+ ├── cost: 0.01
+ └── distribution: test
+
+query T
+EXPLAIN (OPT, MEMO, REDACT) CREATE FUNCTION f() RETURNS STRING LANGUAGE SQL AS $$ SELECT * FROM t128282 WHERE col = 'secret' $$
+----
+memo (optimized, ~8KB, required=[presentation: info:5] [distribution: test])
+ ├── G1: (explain G2 [distribution: test])
+ │    └── [presentation: info:5] [distribution: test]
+ │         ├── best: (explain G2="[distribution: test]" [distribution: test])
+ │         └── cost: 0.03
+ └── G2: (create-function &{‹×› ‹×›
+     ‹×›
+     ‹×›
+     ‹×› [{‹×› ‹×› map[‹×›:0 ‹×›:1 ‹×›:2 ‹×›:3] false 0}] ‹×› ‹×›})
+      ├── [distribution: test]
+      │    ├── best: (create-function &{‹×› ‹×›
+      │    │   ‹×›
+      │    │   ‹×›
+      │    │   ‹×› [{‹×› ‹×› map[‹×›:0 ‹×›:1 ‹×›:2 ‹×›:3] false 0}] ‹×› ‹×›})
+      │    └── cost: 0.01
+      └── []
+           ├── best: (create-function &{‹×› ‹×›
+           │   ‹×›
+           │   ‹×›
+           │   ‹×› [{‹×› ‹×› map[‹×›:0 ‹×›:1 ‹×›:2 ‹×›:3] false 0}] ‹×› ‹×›})
+           └── cost: 0.01
+create-function
+ ├── CREATE FUNCTION f()
+ │     RETURNS STRING
+ │     LANGUAGE SQL
+ │     AS $$SELECT t128282.col FROM test.public.t128282 WHERE col = ‹×›;$$
+ └── dependencies
+      └── t128282 [columns: col]

--- a/pkg/sql/opt/exec/execbuilder/testdata/explain_redact
+++ b/pkg/sql/opt/exec/execbuilder/testdata/explain_redact
@@ -3860,3 +3860,96 @@ project
       └── windows
            └── min [frame="rows from unbounded to unbounded exclude current row"]
                 └── min_1_arg1
+
+# Regression test for #128282: check EXPLAIN (OPT, REDACT) of various CREATE
+# statements.
+
+statement ok
+CREATE TABLE t128282 (col STRING)
+
+query T
+EXPLAIN (OPT, REDACT) CREATE VIEW v AS SELECT * FROM t128282 WHERE col = 'secret'
+----
+create-view .v
+ ├── SELECT t128282.col FROM test.public.t128282 WHERE col = ‹×›
+ ├── columns: col
+ └── dependencies
+      └── t128282 [columns: col]
+
+query T
+EXPLAIN (OPT, VERBOSE, REDACT) CREATE VIEW v AS SELECT * FROM t128282 WHERE col = 'secret'
+----
+create-view .v
+ ├── SELECT t128282.col FROM test.public.t128282 WHERE col = ‹×›
+ ├── columns: col:1
+ ├── dependencies
+ │    └── t128282 [columns: col]
+ ├── cardinality: [0 - 0]
+ ├── volatile, mutations
+ ├── stats: [rows=0]
+ ├── cost: 0.01
+ └── distribution: test
+
+query T
+EXPLAIN (OPT, MEMO, REDACT) CREATE VIEW v AS SELECT * FROM t128282 WHERE col = 'secret'
+----
+memo (optimized, ~8KB, required=[presentation: info:5] [distribution: test])
+ ├── G1: (explain G2 [distribution: test])
+ │    └── [presentation: info:5] [distribution: test]
+ │         ├── best: (explain G2="[distribution: test]" [distribution: test])
+ │         └── cost: 0.03
+ └── G2: (create-view .v)
+      ├── [distribution: test]
+      │    ├── best: (create-view .v)
+      │    └── cost: 0.01
+      └── []
+           ├── best: (create-view .v)
+           └── cost: 0.01
+create-view .v
+ ├── SELECT t128282.col FROM test.public.t128282 WHERE col = ‹×›
+ ├── columns: col
+ └── dependencies
+      └── t128282 [columns: col]
+
+query T
+EXPLAIN (OPT, REDACT) CREATE TABLE t (col STRING CHECK (col != 'secret'))
+----
+create-table
+ └── CREATE TABLE t (col STRING, CHECK (col != ‹×›))
+
+query T
+EXPLAIN (OPT, VERBOSE, REDACT) CREATE TABLE t (col STRING CHECK (col != 'secret'))
+----
+create-table
+ ├── CREATE TABLE t (col STRING, CHECK (col != ‹×›))
+ ├── cardinality: [0 - 0]
+ ├── volatile, mutations
+ ├── stats: [rows=0]
+ ├── cost: 0.02
+ └── distribution: test
+
+query T
+EXPLAIN (OPT, MEMO, REDACT) CREATE TABLE t (col STRING CHECK (col != 'secret'))
+----
+memo (optimized, ~3KB, required=[presentation: info:1] [distribution: test])
+ ├── G1: (explain G2 [distribution: test])
+ │    └── [presentation: info:1] [distribution: test]
+ │         ├── best: (explain G2="[distribution: test]" [distribution: test])
+ │         └── cost: 0.04
+ ├── G2: (create-table G3 &{‹×›  ‹×›})
+ │    ├── [distribution: test]
+ │    │    ├── best: (create-table G3="[distribution: test]" &{‹×›  ‹×›})
+ │    │    └── cost: 0.02
+ │    └── []
+ │         ├── best: (create-table G3 &{‹×›  ‹×›})
+ │         └── cost: 0.02
+ ├── G3: (values G4 id=v1)
+ │    ├── [distribution: test]
+ │    │    ├── best: (values G4 id=v1)
+ │    │    └── cost: 0.01
+ │    └── []
+ │         ├── best: (values G4 id=v1)
+ │         └── cost: 0.01
+ └── G4: (scalar-list)
+create-table
+ └── CREATE TABLE t (col STRING, CHECK (col != ‹×›))

--- a/pkg/sql/opt/memo/expr_format.go
+++ b/pkg/sql/opt/memo/expr_format.go
@@ -726,10 +726,19 @@ func (f *ExprFmtCtx) formatRelational(e RelExpr, tp treeprinter.Node) {
 		}
 
 	case *CreateTableExpr:
-		tp.Child(t.Syntax.String())
+		fmtFlags := tree.FmtSimple
+		if f.RedactableValues {
+			fmtFlags = tree.FmtMarkRedactionNode | tree.FmtOmitNameRedaction
+		}
+		tp.Child(tree.AsStringWithFlags(t.Syntax, fmtFlags))
 
 	case *CreateViewExpr:
-		tp.Child(t.ViewQuery)
+		// Match the format flags used to create t.ViewQuery.
+		fmtFlags := tree.FmtParsable | tree.FmtAlwaysQualifyUserDefinedTypeNames
+		if f.RedactableValues {
+			fmtFlags |= tree.FmtMarkRedactionNode | tree.FmtOmitNameRedaction
+		}
+		tp.Child(tree.AsStringWithFlags(t.Syntax.AsSource, fmtFlags))
 
 		f.Buffer.Reset()
 		f.Buffer.WriteString("columns:")
@@ -741,7 +750,11 @@ func (f *ExprFmtCtx) formatRelational(e RelExpr, tp treeprinter.Node) {
 		f.formatDependencies(tp, t.Deps, t.TypeDeps)
 
 	case *CreateFunctionExpr:
-		tp.Child(t.Syntax.String())
+		fmtFlags := tree.FmtSimple
+		if f.RedactableValues {
+			fmtFlags = tree.FmtMarkRedactionNode | tree.FmtOmitNameRedaction
+		}
+		tp.Child(tree.AsStringWithFlags(t.Syntax, fmtFlags))
 		f.formatDependencies(tp, t.Deps, t.TypeDeps)
 
 	case *CreateStatisticsExpr:

